resolwe/flow/models/utils/duplicate.py

```diff
@@ -0,0 +1,301 @@
+            new_obj = obj_processor(new_obj)
+
+        new_objects.append(new_obj)
+
+    try:
+        # Add another atomic block to avoid corupting the main one.
+        with transaction.atomic():
+            model.objects.bulk_create(new_objects)
+    except IntegrityError:
+        # Probably a slug collision occured, try to create objects one by one.
+        for obj in new_objects:
+            obj.slug = None
+            obj.save()
+
+    for old, new in zip(objects, new_objects):
+        new.created = old.created
+        new.duplicated = timezone.now()
+
+    model.objects.bulk_update(new_objects, ['created', 'duplicated'])
+
+    pk_mapping = {old.pk: new.pk for old, new in zip(objects, new_objects)}
+    _bulk_copy_permissions(content_type, pk_mapping)
+    _bulk_assign_contributor_permission(content_type, pk_mapping.values(), contributor)
+
+    return new_objects
+
+
+@transaction.atomic
+def bulk_duplicate(
+    collections=None,
+    entities=None,
+    data=None,
+    contributor=None,
+    inherit_collection=False,
+    inherit_entity=False,
+    name_prefix=None,
+):
+    """Rewire data inputs, dependencies and save models."""
+    # Prevent circular import.
+    from resolwe.flow.models import (
+        Collection,
+        Data,
+        DataDependency,
+        DataMigrationHistory,
+        Entity,
+    )
+
+    name_prefix = name_prefix or "Copy of"
+
+    if contributor is None:
+        raise ValueError("Contributor must be give.")
+
+    specified = [model for model in [collections, entities, data] if model is not None]
+    if len(specified) != 1:
+        raise ValueError(
+            "Exactly one of 'collections', 'entites', and 'data' attributes must be given."
+        )
+
+    collection_mapping, entity_mapping = {}, {}
+
+    if collections is not None:
+        new_collections = copy_objects(collections, contributor, name_prefix)
+
+        collection_mapping = {
+            old.pk: new for old, new in zip(collections, new_collections)
+        }
+
+        entities = Entity.objects.filter(collection__in=collections)
+        data = Data.objects.filter(collection__in=collections)
+
+        inherit_collection = True
+
+    if entities is not None:
+        entity_processor = process_entity(inherit_collection, collection_mapping)
+        new_entities = copy_objects(
+            entities, contributor, name_prefix, entity_processor
+        )
+
+        entity_mapping = {old.pk: new for old, new in zip(entities, new_entities)}
+
+        entity_data_qs = Data.objects.filter(entity__in=entities)
+        if data is None:
+            data = entity_data_qs
+        else:
+            data = data.union(entity_data_qs)
+
+        inherit_entity = True
+
+    if data is not None:
+        if data.exclude(status__in=[Data.STATUS_DONE, Data.STATUS_ERROR]).exists():
+            raise ValidationError(
+                "Data object must have done or error status to be duplicated."
+            )
```

Comment:
Is this not annoying for the end user? He should wait for all objects in the collection to be processed before being able to duplicate it. 

How about just excluding data objects in processing from the duplication? But then again user will probably expect all objects to be duplicated.